katsdptelstate/telescope_state.py

```diff
@@ -287,6 +306,38 @@ def clear(self):
         """
         return self._backend.clear()
 
+    def _check_immutable_change(self, key, old_enc, new_enc, new):
+        """Check if an immutable is being changed to the same value.
+
+        If not, raise :exc:`ImmutableKeyError`, otherwise just log a message.
+        This is intended to be called by subclasses.
+
+        Parameters
+        ----------
+        key : str
+            Human-readable version of the key
+        old_enc : bytes
+            Previous value, encoded
+        new_enc : bytes
+            New value, encoded
+        new
+            New value, prior to encoding
+        """
+        try:
+            if not equal_encoded_values(new_enc, old_enc):
+                raise ImmutableKeyError(
+                    'Attempt to change value of immutable key {} from '
+                    '{!r} to {!r}'.format(key, decode_value(old_enc), new))
+            else:
+                logger.info('Attribute {} updated with the same value'
+                            .format(key))
+                return True
+        except DecodeError as error:
+            raise ImmutableKeyError(
+                'Attempt to set value of immutable key {} to {!r} but '
+                'failed to decode the previous value to compare: {}'
+                .format(key, new, error))
+
     def add(self, key, value, ts=None, immutable=False, encoding=ENCODING_DEFAULT):
```

Comment:
Frankly I'd like to go the opposite direction and deprecate the `immutable` argument. In practice `add` with `immutable=True` isn't used so much any more now that we have `__setitem__`; and for the asynchronous telstate I'm planning to add a `set` method for setting immutables because you can't easily have an asynchronous `__setitem__`. I find it much clearer to have a different method for each type, which takes a sensible set of arguments for that type, rather than a God method that has lots of parameters to try to cater for every possible use and then have to worry about the interactions.

Reply:
I hear your pain here. Now is a good time for API cleanups, since `katsdptelstate` becomes async, Python 3 only, etc.